lib/bitcoin/SlidingWindowQuantileCalculator.ts

```diff
@@ -0,0 +1,307 @@
+    mongoServerUrl: string,
+    database?: string
+    ) {
+    this.valueApproximator = new ValueApproximator(approximation, maxValue);
+    this.frequencyVectorSize = 1 + this.valueApproximator.getMaximumNormalizedValue();
+    this.slidingWindow = new Array<FrequencyVector>();
+    this.frequencyVectorAggregated = new Array(this.frequencyVectorSize);
+    this.mongoStore = new SlidingWindowQuantileMongoStore(mongoServerUrl, database);
+    if (this.quantileParameter < 0 || this.quantileParameter > 1) {
+      throw Error(`Invalid quantile measure ${quantileParameter}`);
+    }
+  }
+
+  /** Initialize self from state stored in mongo store */
+  public async initialize (): Promise<void> {
+    await this.mongoStore.initialize();
+    const firstBatchId = await this.mongoStore.getFirstBatchId();
+    const lastBatchId = await this.mongoStore.getLastBatchId();
+
+    if (firstBatchId) {
+      for (let batchId = firstBatchId ; batchId <= lastBatchId! ; batchId++) {
+        const quantileInfo = (await this.mongoStore.get(batchId))!;
+
+        this.historicalQuantiles.set(batchId, quantileInfo.quantile);
+
+        if (batchId + this.size > lastBatchId!) {
+          const batchFrequencyVector = runLengthDecode(quantileInfo.batchFreqVector);
+          this.slidingWindow.push(batchFrequencyVector);
+
+          for (let i = 0 ; i < this.frequencyVectorSize ; i++) {
+            this.frequencyVectorAggregated[i] += batchFrequencyVector[i];
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Add a new batch of values to the sliding window. Each batch is
+   * identified using a batchId which can be used to retrieve historical
+   * quantiles as of a particular batchId.
+   */
+  public async add (batchId: number, batch: number[]): Promise<void> {
+
+    // Our historical quantiles storage logic relies on batchIds being
+    // consecutive numbers: explicitly check this fact.
+    if (this.prevBatchId) {
+      if (this.prevBatchId + 1 !== batchId) {
+        throw Error(`Quantile calculator batchIds not sequential`);
+      }
+    }
+    this.prevBatchId = batchId;
+
+    const batchFrequencyVector = new Array(this.frequencyVectorSize);
+
+    for (let value of batch) {
+      const normalizedValue = this.valueApproximator.getNormalizedValue(value);
+      batchFrequencyVector[normalizedValue]++;
+    }
+
+    this.slidingWindow.push(batchFrequencyVector);
+
+    for (let i = 0 ; i < this.frequencyVectorSize ; i++) {
+      this.frequencyVectorAggregated[i] += batchFrequencyVector[i];
+    }
+
+    if (this.slidingWindow.length > this.size) {
+      await this.deleteLast();
+    }
+
+    // calculate and materialize the quantile as of this batchId
+    const quantile = this.calculateCurrentQuantile();
+    this.historicalQuantiles.set(batchId, quantile);
+
+    // store it into mongo store
+    const quantileInfo = {
+      batchId,
+      quantile,
+      batchFreqVector: runLengthEncode(batchFrequencyVector)
```

Comment:
Design sanity: looks like only the "recent" `batchFreqVector` will be needed (for calculating the next quantile)? If that's the case, it doesn't look like there is code for deleting those no-longer-needed batchFrequency vectors from MongoDB. Which leads to the next question: maybe it is fine having to recompute some quantiles when the service restarts and thus we don't store the `batchFreqVector`?

Reply:
We discussed this at length in the last meeting. You are correct that only the recent "batchFreqVector" values are needed. However, with the run length compression and value approximation, batchFreqVector values are really small strings, so I believe it is not worth the coding complexity to remove them. 

Recomputing quantiles when service restarts is probably not a good idea: we want windows to be very large to keep proof-of-fee values smooth; further, this would complicate the bitcoinprocessor code - the recomputation would be doing everything the regular scan is doing but storing the sidetree transactions in the transaction store.

Reply:
I am going to mark this as resolved - please feel free to revive it if necessary.